extensions/amp-nested-menu/validator-amp-nested-menu.protoascii

```diff
@@ -0,0 +1,211 @@
+#
+# Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the license.
+#
+tags: {  # amp-nested-menu
+  html_format: AMP
+  tag_name: "SCRIPT"
+  extension_spec: {
+    name: "amp-nested-menu"
+    version: "0.1"
+    version: "latest"
+    requires_usage: NONE
+  }
+  attr_lists: "common-extension-attrs"
+}
+tags: {  # amp-nested-menu
+  html_format: AMP
+  tag_name: "AMP-NESTED-MENU"
+  # nested menu can be lazy loaded by sidebar.
+  requires_extension: "amp-sidebar"
+  mandatory_ancestor: "AMP-SIDEBAR"
+  attrs: {
+    name: "side"
+    value: "left"
+    value: "right"
+  }
+  attr_lists: "extended-amp-global"
+  amp_layout: {
+    supported_layouts: FILL
+  }
+  descendant_tag_list: "amp-nested-menu-allowed-descendants"
+  spec_url: "https://amp.dev/documentation/components/amp-nested-menu"
+}
+tags: {  # <amp-nested-menu> <div amp-nested-submenu(-open/-close)>
+  html_format: AMP
+  tag_name: "DIV"
+  spec_name: "AMP-NESTED-MENU DIV"
+  mandatory_ancestor: "AMP-NESTED-MENU"
+  # submenus inside accordion do not display correctly due to styling of <amp-accordion>.
+  disallowed_ancestor: "AMP-ACCORDION"
+  attrs: {
+    name: "amp-nested-submenu"
+    mandatory_oneof: "['amp-nested-submenu', 'amp-nested-submenu-close', 'amp-nested-submenu-open']"
+  }
+  attrs: {
+    name: "amp-nested-submenu-close"
+    mandatory_oneof: "['amp-nested-submenu', 'amp-nested-submenu-close', 'amp-nested-submenu-open']"
+  }
+  attrs: {
+    name: "amp-nested-submenu-open"
+    mandatory_oneof: "['amp-nested-submenu', 'amp-nested-submenu-close', 'amp-nested-submenu-open']"
+  }
+}
+tags: {  # <amp-nested-menu> <button amp-nested-submenu-open/close>
```

Comment:
Is there a better way to express this rule? 

> within `amp-nested-menu`, allowed elements can contain either one but not both of these attributes: 'amp-nested-submenu-close', 'amp-nested-submenu-open'

Where allowed elements are: heading tags, buttons, or spans. 

Right now, we individually list and specify them, but it would probably be easier if we had one rule and a list of tags to which they applied.

Reply:
Yes. Use `attr_lists`, defined [here](https://github.com/ampproject/amphtml/blob/master/validator/validator.proto#L184). Example [usage](https://github.com/ampproject/amphtml/blob/master/validator/validator-main.protoascii#L183).

Code:

```
attr_lists {
  name: "amp-nested-menu-attrs"
  attrs: {
    name: "amp-nested-submenu"
    mandatory_oneof: "['amp-nested-submenu', 'amp-nested-submenu-close', 'amp-nested-submenu-open']"
  }
  attrs: {
    name: "amp-nested-submenu-close"
    mandatory_oneof: "['amp-nested-submenu', 'amp-nested-submenu-close', 'amp-nested-submenu-open']"
  }
  attrs: {
    name: "amp-nested-submenu-open"
    mandatory_oneof: "['amp-nested-submenu', 'amp-nested-submenu-close', 'amp-nested-submenu-open']"
  }
}
```

Then for each tag element specify it by name:

```
tags: {
  ...
  attr_lists: "amp-nested-menu-attrs"
  ...
}
````

Reply:
And note my code is wrong since the ones below do not have `amp-nested-submenu`, only the `-open` and `-close`

Reply:
That saves writing out the individual attributes, but I guess we still have to defined the tags one by one?

Reply:
Yes, there are no grouping of TagSpecs which have specific requirements like this on the tag itself. There are grouping of tag names where it serves a specific purpose, like [DescendantTagList](https://github.com/ampproject/amphtml/blob/master/validator/validator.proto#L191).